### PR TITLE
react: Make value returned by useState Readonly

### PR DESCRIPTION
**Pull request by SpLouk**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
  - Not sure what to add as a test for this; this change is intended to make mutation of state impossible
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

State is immutable, so the return value of useState should be wrapped in the `DeepReadonly` utility type which can prevent accidental mutation.

https://github.com/piotrwitek/utility-types/blob/master/src/mapped-types.ts#L396

This will prevent accidental mutation of state. E.g.

```
const [arr, setArr] = useState(['a', 'b']);
arr.push('c')
```

### CODE REVIEW

**Comment by Semigradsky** [comment]
I like this suggestion, but it will break a lot of existing code.

**Review by eps1lon — COMMENTED** [review]
> I like this suggestion, but it will break a lot of existing code.

Yeah I can see that. Just had a case today were a component library was expecting a mutable array (which is the default in TypeScript). In those cases, TypeScript would now throw.

Could you test this on some small/medium+ sized repositories to check how many new issues we create with this change?

**Comment by psalv** [comment]
> I like this suggestion, but it will break a lot of existing code.

It's arguable this code has a high chance of being broken already, they just don't know about it

File: types/react/v17/index.d.ts [hunk 1]
```diff
@@ -918,15 +919,15 @@ declare namespace React {
      * @version 16.8.0
      * @see https://reactjs.org/docs/hooks-reference.html#usestate
      */
-    function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
+    function useState<S>(initialState: S | (() => S)): [DeepReadonly<S>, Dispatch<SetStateAction<S>>];
```
[the remark this answers is not in the document]
Comment (SpLouk) [reply types/react/v17/index.d.ts:922]: I think DeepReadonly would be preferred to prevent mutating nested members of an object.

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react@latest/))
was missing the following properties:
1. unstable_act
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4665ec5e7744a6fc5d57bc0839b37a29811813cd
</p>

**Comment by eps1lon** [comment]
> It's arguable this code has a high chance of being broken already, they just don't know about it

Did you run this on some example codebases? What were patterns that it did reject, that we want to reject and what were patterns we don't necessarily want to reject (e.g. due to the ecosystem not having props as readonly in most cases)?

**Comment by SpLouk** [comment]
Yeah, so this will definitely be a breaking change. I didn't anticipate this but 

```
const [str,...] = useStr(['a', 'b'])

return <Child str={ str } /> // pass the state down to a child
```

Will cause a type error because `Child` is expecting a `string []` not a `readonly string[]`.

I ran our type checker in Faire's codebase locally and got about 20 type errors of this form in a ~100,000 line React/TS codebase.

Now, in a perfect world, all arrays and objects passed as props should be readonly too, but that isn't the case in our codebase and I'm sure most others.

**Comment by eps1lon** [comment]
> Now, in a perfect world, all arrays and objects passed as props should be readonly too, but that isn't the case in our codebase and I'm sure many others.

Same for our design system at Klarna.

Did you catch any actual errors similar to the one explained in the PR description?

**Comment by SpLouk** [comment]
> Did you catch any actual errors similar to the one explained in the PR description?

No, no instances of an object actually being mutated.

**Comment by SpLouk** [comment]
Going to go with an internal solution here.

**Pull request closed without merging**